File: packages/urdu-js/tests/encoding/inpage.test.ts
import { describe, it, expect } from 'vitest'
import { decodeInpage } from '../../src/encoding/inpage/decoder.js'

describe('decodeInpage v1', () => {
  it('empty buffer → no paragraphs', () => {
    expect(decodeInpage(new Uint8Array([]), 'v1').paragraphs).toHaveLength(0)
  })
  it('decodes alef: 0x04 0x81 → ا', () => {
    expect(decodeInpage(new Uint8Array([0x04, 0x81]), 'v1').paragraphs[0]).toBe('\u0627')
  })
  it('decodes پ: 0x04 0x83 → پ', () => {
    expect(decodeInpage(new Uint8Array([0x04, 0x83]), 'v1').paragraphs[0]).toBe('\u067e')
  })
  it('page break 0x0C creates break index', () => {
    const r = decodeInpage(new Uint8Array([0x04, 0x81, 0x0c, 0x04, 0x82]), 'v1')
    expect(r.pageBreakIndices).toHaveLength(1)
    expect(r.paragraphs).toHaveLength(2)
  })
  it('newline 0x0A splits paragraphs', () => {
    const r = decodeInpage(new Uint8Array([0x04, 0x81, 0x0a, 0x04, 0x82]), 'v1')
    expect(r.paragraphs).toHaveLength(2)
  })
  it('ASCII chars pass through', () => {
    const buf = new Uint8Array([0x48, 0x69]) // "Hi"
    expect(decodeInpage(buf, 'v1').paragraphs[0]).toBe('Hi')
  })
})

File: packages/urdu-dart/test/encoding/inpage_test.dart
import 'package:test/test.dart';
import 'dart:typed_data';
import 'package:urdu_tools/urdu_tools.dart';

void main() {
  group('decodeInpage v1', () {
    test('empty buffer -> no paragraphs', () {
      expect(decodeInpage(Uint8List(0), InpageVersion.v1).paragraphs, isEmpty);
    });
    test('decodes alef: 0x04 0x81 -> ا', () {
      expect(
          decodeInpage(Uint8List.fromList([0x04, 0x81]), InpageVersion.v1)
              .paragraphs[0],
          equals('\u0627'));
    });
    test('decodes پ: 0x04 0x83 -> پ', () {
      expect(
          decodeInpage(Uint8List.fromList([0x04, 0x83]), InpageVersion.v1)
              .paragraphs[0],
          equals('\u067e'));
    });
    test('page break 0x0C creates break index', () {
      final r = decodeInpage(
          Uint8List.fromList([0x04, 0x81, 0x0c, 0x04, 0x82]), InpageVersion.v1);
      expect(r.pageBreakIndices, hasLength(1));
      expect(r.paragraphs, hasLength(2));
    });
    test('newline 0x0A splits paragraphs', () {
      final r = decodeInpage(
          Uint8List.fromList([0x04, 0x81, 0x0a, 0x04, 0x82]), InpageVersion.v1);
      expect(r.paragraphs, hasLength(2));
    });
    test('ASCII chars pass through', () {
      final buf = Uint8List.fromList([0x48, 0x69]); // "Hi"
      expect(decodeInpage(buf, InpageVersion.v1).paragraphs[0], equals('Hi'));
    });
    test('space byte 0x20 is preserved', () {
      final r = decodeInpage(
          Uint8List.fromList([0x04, 0x81, 0x20, 0x04, 0x81]), InpageVersion.v1);
      expect(r.paragraphs[0], equals('\u0627 \u0627'));
    });
    test('carriage return 0x0D splits paragraphs', () {
      final r = decodeInpage(
          Uint8List.fromList([0x04, 0x81, 0x0d, 0x04, 0x81]), InpageVersion.v1);
      expect(r.paragraphs, hasLength(2));
    });
    test('unknown byte is skipped without crash', () {
      final r = decodeInpage(
          Uint8List.fromList([0xff, 0x04, 0x81]), InpageVersion.v1);
      expect(r.paragraphs[0], equals('\u0627'));
    });
  });

  group('decodeInpage v3', () {
    test('empty buffer -> no paragraphs', () {
      expect(decodeInpage(Uint8List(0), InpageVersion.v3).paragraphs, isEmpty);
    });
    test('decodes UTF-16LE Urdu codepoint', () {
      final r =
          decodeInpage(Uint8List.fromList([0x27, 0x06]), InpageVersion.v3);
      expect(r.paragraphs[0], equals('\u0627'));
    });
    test('0x000D splits paragraphs in v3', () {
      final r = decodeInpage(
          Uint8List.fromList([0x27, 0x06, 0x0d, 0x00, 0x28, 0x06]),
          InpageVersion.v3);
      expect(r.paragraphs, hasLength(2));
    });
    test('0x000C creates page break in v3', () {
      final r = decodeInpage(
          Uint8List.fromList([0x27, 0x06, 0x0c, 0x00, 0x28, 0x06]),
          InpageVersion.v3);
      expect(r.pageBreakIndices, hasLength(1));
    });
    test('0xFFFF paragraph marker splits in v3', () {
      final r = decodeInpage(
          Uint8List.fromList([0x27, 0x06, 0xff, 0xff, 0x28, 0x06]),
          InpageVersion.v3);
      expect(r.paragraphs, hasLength(2));
    });
  });

  group('decodeInpage auto', () {
    test('high 0x04 density -> v1 decode', () {
      final buf = Uint8List(20)..fillRange(0, 20, 0x04);
      expect(() => decodeInpage(buf, InpageVersion.auto), returnsNormally);
    });
    test('low 0x04 density -> v3 decode produces Urdu text', () {
      final buf = Uint8List.fromList([0x27, 0x06, 0x28, 0x06]);
      expect(decodeInpage(buf, InpageVersion.auto).paragraphs[0],
          equals('\u0627\u0628'));
    });
  });
}
